app/src/main/java/org/mozilla/focus/permission/PermissionHandler.java

```diff
@@ -0,0 +1,152 @@
+package org.mozilla.focus.permission;
+
+import android.Manifest;
+import android.app.Activity;
+import android.content.Context;
+import android.content.DialogInterface;
+import android.content.Intent;
+import android.content.SharedPreferences;
+import android.content.pm.PackageManager;
+import android.os.Build;
+import android.os.Bundle;
+import android.preference.PreferenceManager;
+import android.support.annotation.NonNull;
+import android.support.design.widget.Snackbar;
+import android.support.v4.app.ActivityCompat;
+import android.support.v4.content.ContextCompat;
+import android.support.v7.app.AlertDialog;
+import android.view.View;
+
+import org.mozilla.focus.R;
+import org.mozilla.focus.utils.IntentUtils;
+
+import java.io.Serializable;
+
+public class PermissionHandler {
+
+    private static final int REQUEST_SETTINGS = 39528;
+    private static final String PERMISSION_KEY = "HANDLER_PERMISSION_KEY";
+    private static final String ACTION_ID_KEY = "HANDLER_ACTION_ID_KEY";
+    private static final String PARAMS_KEY = "HANDLER_PARAMS_KEY";
+    private static final String PERMISSION_PREFIX = "PERM_";
+
+
+    private PermissionHandle permissionHandle;
+    private String permission;
+    private int actionId;
+    private Serializable params;
+
+    public PermissionHandler(PermissionHandle permissionHandle) {
+        this.permissionHandle = permissionHandle;
+    }
+
+    public void tryAction(final Activity activity, final String permission, final int actionId, final Serializable params) {
+        if (PackageManager.PERMISSION_GRANTED == ContextCompat.checkSelfPermission(activity, permission)) {
```

Comment:
Try to reduce if/else depth level, maybe "early return" would be enough.

Reply:
Done!